### PR TITLE
Create PEL with inventory callouts

### PR DESCRIPTION
**Pull request by branupama**
This commit adds the changes for implementing async PEL creation API, with inventory callout’s information provided by the caller.

### CODE REVIEW

File: src/event_logger.cpp [hunk 1]
```diff
@@ -60,17 +60,90 @@ void EventLogger::createAsyncPelWithInventoryCallout(
...
+        std::string l_message = m_errorMsgMap.at(i_errorType);
+
+        std::string l_severity =
+            (m_severityMap.find(i_severity) != m_severityMap.end()
+                 ? m_severityMap.at(i_severity)
+                 : m_severityMap.at(types::SeverityType::Error));
+
+        std::string l_description("Description field missing in user data");
+
+        if (i_userData1 && ((*i_userData1).first.compare("DESCRIPTION") ==
```
Comment (branupama) [inline src/event_logger.cpp:90]: if caller not passing "DESCRIPTION" key, instead if they pass any other value as key.

should we add "DESCRIPTION" key and value as "Description field missing in user data" along with caller passed
key:value to PEL ?

File: src/event_logger.cpp [hunk 2]
```diff
@@ -60,17 +60,90 @@ void EventLogger::createAsyncPelWithInventoryCallout(
...
+        }
+
+        if (m_errorMsgMap.find(i_errorType) == m_errorMsgMap.end())
+        {
+            logging::logMessage(
+                "PEL error type not found in the error message map.");
+            return;
+        }
+
+        std::string l_message = m_errorMsgMap.at(i_errorType);
```
Comment (SunnySrivastava1984) [inline src/event_logger.cpp:81]: Can we avoid string copy here?

File: src/event_logger.cpp [hunk 3]
```diff
@@ -60,17 +60,90 @@ void EventLogger::createAsyncPelWithInventoryCallout(
...
+            l_description = (*i_userData1).second;
+        }
+
+        sd_bus* l_sdBus = nullptr;
+        sd_bus_default(&l_sdBus);
+
+        types::InventoryCalloutData l_invCallout = i_callouts[0];
+        types::CalloutPriority l_priorityEnum = get<1>(l_invCallout);
+
+        std::string l_priority =
```
Comment (SunnySrivastava1984) [inline src/event_logger.cpp:102]: Unless not required, please use const reference?

File: src/event_logger.cpp (hunk 1 above)
Comment (SunnySrivastava1984) [reply src/event_logger.cpp:90]: I don't see a downside to that.

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]
Need a better commit message.

File: src/event_logger.cpp (hunk 2 above)
Comment (branupama) [reply src/event_logger.cpp:81]: Done

File: src/event_logger.cpp [hunk 4]
```diff
@@ -60,17 +60,90 @@ void EventLogger::createAsyncPelWithInventoryCallout(
...
+        if (m_errorMsgMap.find(i_errorType) == m_errorMsgMap.end())
+        {
+            logging::logMessage(
+                "PEL error type not found in the error message map.");
+            return;
+        }
+
+        std::string l_message = m_errorMsgMap.at(i_errorType);
+
+        std::string l_severity =
```
[the remark this answers is not in the document]
Comment (branupama) [reply src/event_logger.cpp:83]: Done

File: src/event_logger.cpp (hunk 3 above)
Comment (branupama) [reply src/event_logger.cpp:102]: Done

File: src/event_logger.cpp [hunk 5]
```diff
@@ -60,17 +60,90 @@ void EventLogger::createAsyncPelWithInventoryCallout(
...
+        auto l_rc = -1;
+        if (!i_userData2)
+        {
+            l_rc = sd_bus_call_method_async(
+                l_sdBus, NULL, constants::eventLoggingServiceName,
+                constants::eventLoggingObjectPath,
+                constants::eventLoggingInterface, "Create", NULL, NULL,
+                "ssa{ss}", l_message.c_str(), l_severity.c_str(), 6, "FileName",
+                i_fileName.c_str(), "FunctionName", i_funcName.c_str(),
+                "InternalRc", i_internalRc.c_str(), "DESCRIPTION",
```
Comment (branupama) [inline src/event_logger.cpp:116]: is should be "InternalReturnCode" or "InternalReasonCode" ? and
is "FileName" or "File Name" ?

File: src/event_logger.cpp [hunk 6]
```diff
@@ -60,17 +60,90 @@ void EventLogger::createAsyncPelWithInventoryCallout(
...
+        if (i_userData1 && ((*i_userData1).first.compare("DESCRIPTION") ==
+                            constants::STR_CMP_SUCCESS))
+        {
+            l_description = (*i_userData1).second;
+        }
+
+        sd_bus* l_sdBus = nullptr;
+        sd_bus_default(&l_sdBus);
+
+        const types::InventoryCalloutData& l_invCallout = i_callouts[0];
```
Comment (jinuthomas) [inline src/event_logger.cpp:99]: shouldn't this be in a loop ?

File: src/event_logger.cpp [hunk 7]
```diff
@@ -60,17 +60,90 @@ void EventLogger::createAsyncPelWithInventoryCallout(
...
+        const std::string& l_message = m_errorMsgMap.at(i_errorType);
+
+        const std::string& l_severity =
+            (m_severityMap.find(i_severity) != m_severityMap.end()
+                 ? m_severityMap.at(i_severity)
+                 : m_severityMap.at(types::SeverityType::Error));
+
+        std::string l_description("Description field missing in user data");
+
+        if (i_userData1 && ((*i_userData1).first.compare("DESCRIPTION") ==
```
Comment (jinuthomas) [inline src/event_logger.cpp:90]: If userdata's first string is DESCRIPTION also for the userdata2 , then why do we need it as a parameter?
Also is DESCRIPTION mandatory for and error log ?

File: src/event_logger.cpp (hunk 5 above)
Comment (jinuthomas) [reply src/event_logger.cpp:116]: Did not understand why this question

File: src/event_logger.cpp [hunk 8]
```diff
@@ -60,17 +60,90 @@ void EventLogger::createAsyncPelWithInventoryCallout(
...
+        if (l_rc < 0)
+        {
+            logging::logMessage(
+                "Error calling sd_bus_call_method_async, Message = " +
+                std::string(strerror(-l_rc)));
+        }
+    }
+    catch (const sdbusplus::exception::SdBusError& l_ex)
+    {
+        logging::logMessage(l_ex.what());
```
Comment (jinuthomas) [inline src/event_logger.cpp:145]: just remind me please on why this try catch , what functions will be giving exceptions in this case.

**Review by jinuthomas — CHANGES_REQUESTED** [review]

File: src/event_logger.cpp (hunk 6 above)
Comment (branupama) [reply src/event_logger.cpp:99]: as currently only one INVENTORY_PATH callout is supported by calling 'create' API from 'Logging' service. we are handling single callout for now.

Added todo in future support for multiple callouts.

File: src/event_logger.cpp (hunk 8 above)
Comment (branupama) [reply src/event_logger.cpp:145]: if any exception occurs in this API, instead throwing back to caller its handled here by printing the error and returning caller quietly.
Updated the exception.

**Review by SunnySrivastava1984 — APPROVED** [review]

File: include/event_logger.hpp [hunk 9]
```diff
@@ -40,21 +40,27 @@ class EventLogger
      * @param[in] i_fileName - File name.
      * @param[in] i_funcName - Function name.
      * @param[in] i_internalRc - Internal return code.
-     * @param[in] i_userData1 - Additional user data [optional].
-     * @param[in] i_userData2 - Additional user data [optional].
+     * @param[in] i_description - Additional user data value, 'DESCRIPTION' as
```
Comment (jinuthomas) [inline include/event_logger.hpp:43]: why mention key in doxygen

**Review by jinuthomas — CHANGES_REQUESTED** [review]

**Pull request merged**